### PR TITLE
[claude][skill] distribute triaging sub-skill

### PR DESCRIPTION
**Pull request by anshul-si**
tested in pytorch/ciforge
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #180401

### CODE REVIEW

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+# Distributed Triage Rubric
+
+Detailed guidance for sub-triaging issues in the `oncall: distributed` queue. Use this rubric alongside [distributed-labels.json](distributed-labels.json) when classifying issues.
+
+---
+
+## 1. Sub-Oncall Routing Rules
+
+Route each issue to exactly ONE sub-oncall. When in doubt, default to `oncall: distributed parallelisms`.
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:9]: Thought: Should we have the bot comment "Why" it decides to route an issue a certain way?

e.g.,

[ptd-triage-bot] Routing issue to `oncall: distributed parallelisms` because the issue appears to pertain to FSDP as per user's description. 

These comments could be examined by future bots -- e.g., the second round of triaging to assign module/priority

File: .claude/skills/triaging-issues/distributed-triage/distributed-labels.json [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    {
+      "name": "module: context parallel",
+      "description": "Context parallelism"
+    },
+    {
+      "name": "module: activation checkpointing",
+      "description": "Activation checkpointing (when used with distributed)"
+    },
+    {
+      "name": "module: multi-gpu",
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-labels.json:70]: I am genuinely not sure how valuable is this label...

File: .claude/skills/triaging-issues/distributed-triage/distributed-labels.json [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    {
+      "name": "feature",
+      "description": "Wholly new functionality that does not exist today"
+    },
+    {
+      "name": "enhancement",
+      "description": "Improvement to something that already works"
+    },
+    {
+      "name": "function request",
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-labels.json:98]: Is this really useful? Not sure where one would draw the line between this and enhancement.

File: .claude/skills/triaging-issues/distributed-triage/distributed-labels.json [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    {
+      "name": "triage review",
+      "description": "Needs human review"
+    },
+    {
+      "name": "needs reproduction",
+      "description": "Needs a minimal reproduction script"
+    },
+    {
+      "name": "ptd-bot-triaged",
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-labels.json:114]: repeated. Already present in line 86

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+### `oncall: distributed parallelisms` (default)
+
+Covers the parallel training APIs that users interact with directly:
+- **FSDP** (FSDP1 and FSDP2): `FullyShardedDataParallel`, `fully_shard`, sharding strategies, mixed precision
+- **DDP**: `DistributedDataParallel`, gradient synchronization, parameter broadcasting
+- **DTensor**: `distribute_tensor`, placements (`Shard`, `Replicate`), tensor redistribution
+- **DeviceMesh**: `init_device_mesh`, mesh dimensions, multi-dimensional parallelism
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:17]: I think DeviceMesh falls under distributed infra

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+- **Tensor Parallel**: column/row parallel, sequence parallel
+- **Context Parallel**: context parallelism for long sequences
+- **Pipeline Parallel**: `PipelineSchedule`, `PipelineStage`, pipeline schedules
+- **Symmetric Memory**: `SymmetricMemory`, `symm_mem`
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:21]: Same for Symmetric Memory

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+Covers saving/loading distributed model state:
+- **DCP** (Distributed Checkpoint): `torch.distributed.checkpoint`, `save`, `load`, `state_dict`
+- **State dict utilities**: `get_model_state_dict`, `get_optimizer_state_dict`, `set_model_state_dict`
+- **Checkpoint format**: file system planner, HDF5, resharding across different world sizes
+- **Async checkpointing**: non-blocking checkpoint operations
+
+### Routing Decision Tree
+
+```
+Is the issue about saving/loading distributed checkpoints or state_dicts?
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:45]: Instead of being so brief here, can we have it consult the oncall charters listed in previous section to figure out the routing?

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+Is the issue about the communication layer (process groups, collectives, backends,
+   NCCL errors, torchrun/elastic, RPC)?
+  YES → oncall: distributed infra
+
+Otherwise → oncall: distributed parallelisms
+```
+
+### Edge Cases
+
+- **NCCL timeout during FSDP training**: Route to `oncall: distributed infra`. The NCCL timeout is the bug, FSDP is just the user context.
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:57]: I am not sure about this one. NCCL timeout could be a symptom of a bug elsewhere. This item might merit more examples to guide the agent. We could start with dist infra ...

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+Detailed guidance for sub-triaging issues in the `oncall: distributed` queue. Use this rubric alongside [distributed-labels.json](distributed-labels.json) when classifying issues.
+
+---
+
+## 1. Sub-Oncall Routing Rules
+
+Route each issue to exactly ONE sub-oncall. When in doubt, default to `oncall: distributed parallelisms`.
+
+### `oncall: distributed parallelisms` (default)
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:11]: Should you also mention what module tags fall under the purview of each oncall?

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+- **Confusions**: Only label this if the issue is about activation checkpointing specifically, not about distributed checkpoints (saving model state). If it's "checkpoint saving", that's `oncall: distributed checkpointing`.
+
+---
+
+## 3. Confidence Calibration
+
+### HIGH Confidence Examples
+
+- Issue title: "FSDP2 crashes with `fully_shard` when using `MixedPrecision`" → `module: fsdp` (explicit API mention)
+- Stack trace shows `torch/distributed/fsdp/fully_sharded_data_parallel.py:423` → `module: fsdp`
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:144]: Relying on stack traces could get dicey if the error is lower down the stack, but fsdp is in an earlier stack frame.

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+| `mcp__github__add_issue_comment` | Add comment (only for reproduction requests or mislabel flags) |
+| `mcp__github__search_issues` | Find similar issues for context |
+
+---
+
+## Distributed Triage Steps
+
+### 0) Already Triaged by Human?
+
+Check if the issue already has a distributed `module:` label from [distributed-labels.json](distributed-labels.json). The distributed module labels are:
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/SKILL.md:48]: do you need to re-list the module labels here? Better to list in just one place?

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+| Sub-Oncall Label | When to Apply |
+|-----------------|---------------|
+| `oncall: distributed parallelisms` | FSDP, DDP, DTensor, DeviceMesh, tensor parallel, context parallel, pipeline parallel, symmetric memory. **This is the default** when unsure. |
+| `oncall: distributed infra` | c10d, process groups, collectives, NCCL/Gloo/MPI backends, elastic/torchrun, RPC, stores, distributed tools |
+| `oncall: distributed checkpointing` | Distributed checkpoint save/load, DCP, state_dict utilities, async checkpointing |
+
+Use the routing decision tree and edge cases in [distributed-rubric.md](distributed-rubric.md) Section 1 to determine the correct sub-oncall.
+
+**After routing to `oncall: distributed infra` or `oncall: distributed checkpointing`:**
+- Add `ptd-bot-triaged`
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/SKILL.md:86]: I don't think you should add this label if the bot just routed to an oncall? 

IMO `ptd-bot-triaged` should be added only when the issue has been triaged to its last-mile module.

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+- `function request` — a new function or new arguments/modes for an existing function
+
+Most distributed issues are bug reports — do not add a type label for bugs. If the issue says the operation "currently works" or "falls back to" a slower path, that is `enhancement`, not `feature`. If the enhancement is about performance, also add `module: performance`.
+
+### 5) High Priority — REQUIRES HUMAN REVIEW
+
+**CRITICAL:** If you believe an issue is high priority, you MUST:
+1. Add `triage review` label and do NOT add `ptd-bot-triaged`
+
+Do NOT directly add `high priority` without human confirmation.
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/SKILL.md:123]: Why not? The bot can flag as high priority + triage review, right?

Comment (aditvenk) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md:123]: The MyClaw message to oncall should also list these issues as most important for the oncall to check out.

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+**DO NOT:**
+- Close issues (only the PT-level bot or humans close issues)
+- Remove existing labels — only add labels
+- Remove `oncall: distributed` — it stays even if the issue is mislabeled
+- Remove `oncall: pt2` — if already present, keep it
+- Add labels not in [distributed-labels.json](distributed-labels.json)
+- Add comments to issues except when using the templates in Step 1 (mislabel) or Step 6 (reproduction)
+- Assign issues to users
+- Add `high priority` directly — use `triage review` and let humans decide
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/SKILL.md:158]: this is the one that feels unnecessary to me

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+**DO:**
+- Be conservative — when in doubt, add `triage review` for human attention
+- Add `ptd-bot-triaged` only when you are confident in the classification (HIGH confidence)
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/SKILL.md:162]: Once module has been classified, it should add this triaged bot?

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+- The issue is a feature request (no repro needed)
+- A multi-node script is provided (that counts as reproduction even if you can't run it locally)
+
+---
+
+## Constraints
+
+**DO NOT:**
+- Close issues (only the PT-level bot or humans close issues)
+- Remove existing labels — only add labels
```
Comment (aditvenk) [inline .claude/skills/triaging-issues/distributed-triage/SKILL.md:152]: worth calling out to not remote `bot-triaged` label as well -- as it is likely to come up all the time

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md (hunk 9 above)
Comment (anshul-si) [reply .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:11]: I agree, wanted to confirm if we should have used modules to determine oncalls before I did this

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md (hunk 12 above)
Comment (anshul-si) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md:85]: should there be some tag then that says bot sent it to specific oncall. i think there should be some way to distinguish the presence of the second bot so that we can add lessons based on its mistakes

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md (hunk 13 above)
Comment (anshul-si) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md:120]: I just copied the exact high priority handling that pt2 has for its triaging. I'll check why this is handled this way

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md (hunk 15 above)
Comment (anshul-si) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md]: ok my reasoning for this is to help improve the bot. I think the bot should add modules even if not highly confident. then we can add lessons based on correctness so help the bot classify less confidence issues in the future. in these cases bot will send to triage review which seems contradictory to also added triaged-bot?

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md (hunk 12 above)
Comment (aditvenk) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md:85]: Fair point -- the confusion comes from our choice of using `ptd-bot-triaged` for the case where the bot does the last mile triaging -- it should really just tag "triaged" in that case. 

Let's use `ptd-bot-triaged` for all cases where the bot has taken some action. 

1) route to oncall? -> `ptd-bot-triaged`
2) route to module + priority? -> `ptd-bot-triaged` (for now); later `triaged` as well, once we have confidence in the bot
3) unable to route module? -> `ptd-bot-triaged` + `triage-review`

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md (hunk 15 above)
Comment (aditvenk) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md]: `ptd-bot-triaged` + `triage-review` is okay. It means the bot took a shot at it, but needs human to do the rest

File: .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md (hunk 1 above)
Comment (anshul-si) [reply .claude/skills/triaging-issues/distributed-triage/distributed-rubric.md:9]: The /data/users/anshulsi/pytorch/.github/workflows/claude-distributed-triage.yml script already uploads the full bot conversation output to an S3 bucket. This just unnecessarily adds extra noise to the issue conversation

File: .claude/skills/triaging-issues/distributed-triage/SKILL.md (hunk 14 above)
Comment (anshul-si) [reply .claude/skills/triaging-issues/distributed-triage/SKILL.md:156]: for now i will keep it this way in line with what pt2 triage does until i understand reasoning and why it doesn't need to apply to us

**Review by aditvenk — APPROVED** [review]
